### PR TITLE
fix ajv

### PR DESCRIPTION
**Pull request by bluebill1049**

### CODE REVIEW

**Comment by jorisre** [comment]
Thanks @bluebill1049 ! This will fix umd bundle :)

**Comment by bluebill1049** [comment]
> Thanks @bluebill1049 ! This will fix umd bundle :)

doesn't looks like it's working... :(

**Comment by bluebill1049** [comment]
Screen Shot 2020-09-18 at 4 15 03 pm

I think the issue is import Ajv, i will remove the import and should should construct the schema.

**Comment by bluebill1049** [comment]
@TheDSCPL FYI

File: src/jsonSchema.ts [hunk 1]
```diff
@@ -1,8 +1,8 @@
 import * as Ajv from 'ajv';
 import { appendErrors, Resolver, transformToNestObject } from 'react-hook-form';
-import type { ErrorObject, Options, ValidateFunction } from 'ajv';
-import type { JSONSchema4, JSONSchema6, JSONSchema7 } from 'json-schema';
-import type {
+import { ErrorObject, Options, ValidateFunction } from 'ajv';
```
Comment (TheDSCPL) [inline src/jsonSchema.ts:3]: Why did you delete the "type"? It was correct

File: src/jsonSchema.ts [hunk 2]
```diff
@@ -1,8 +1,8 @@
 import * as Ajv from 'ajv';
```
Comment (TheDSCPL) [inline src/jsonSchema.ts:1]: When you add `esModuleInterop: true` to the `tsconfig.json`, `* as ` should be removed and all the type errors will disapear.

File: src/jsonSchema.ts [hunk 3]
```diff
@@ -79,7 +79,8 @@ export const jsonSchemaResolver = <TFieldValues extends Record<string, any>>(
     // compile only if necessary. otherwise return the memoized validation function
     if (!validate || allErrors !== validateAllFieldCriteria) {
       allErrors = validateAllFieldCriteria;
-      validate = new Ajv({
+      // @ts-ignore
+      validate = new Ajv.default({
```
Comment (TheDSCPL) [inline src/jsonSchema.ts:83]: This shouldn't be done. TS was made by very clever people who know how the modules systems work. If it's an error, it won't work on runtime. In this case changing the import and changing this line back to how it was will remove the error.

File: src/jsonSchema.ts [hunk 4]
```diff
@@ -88,6 +89,7 @@ export const jsonSchemaResolver = <TFieldValues extends Record<string, any>>(
       }).compile(validationSchema);
     }
 
+    // @ts-ignore
```
Comment (TheDSCPL) [inline src/jsonSchema.ts:92]: What was the error here? Btw for future reference using `@ts-expect-error` is recommended instead of `@ts-ignore`. But even that should seldom be used.

File: src/jsonSchema.ts (hunk 1 above)
Comment (bluebill1049) [reply src/jsonSchema.ts:3]: Screen Shot 2020-09-18 at 4 25 07 pm

File: src/jsonSchema.ts (hunk 2 above)
Comment (bluebill1049) [reply src/jsonSchema.ts:1]: Screen Shot 2020-09-18 at 4 25 32 pm

File: src/jsonSchema.ts (hunk 3 above)
Comment (bluebill1049) [reply src/jsonSchema.ts:83]: i build it and failed without it.

File: src/jsonSchema.ts (hunk 4 above)
Comment (bluebill1049) [reply src/jsonSchema.ts:92]: at this stage, I am trying to get it working first.

**Comment by bluebill1049** [comment]
@TheDSCPL I am going to try to pull Ajv out.
Screen Shot 2020-09-18 at 4 26 26 pm

**Comment by TheDSCPL** [comment]
What do you mean? Delete it?

**Comment by bluebill1049** [comment]
> What do you mean? Delete it?

no, move the new Ajv out from the resolver, so we don't need to import Ajv. that's the root issue.

**Comment by bluebill1049** [comment]
> What do you mean? Delete it?

unless you have a better strategy to fix it. as it's probably a blocker right now.

**Comment by TheDSCPL** [comment]
> no, move the new Ajv out from the resolver, so we don't need to import Ajv. that's the root issue.

I think that will have to the next step. As I've said before, I think separating the resolver implementations into different sub-packages under `@hookform` will be the only effective solution for the users to not install 3 extra packages (and their respective dependencies, dependencies of dependencies, ...) for just using one of the implementations.

Anyway, that is not the issue here, though.

**Comment by jorisre** [comment]
No sure about this solution but to make it works, you can add `external: ['ajv']` to rollup config : 

```js
  const config = {
    input: options.input,
    external: ['ajv'], // <= It works !
    output: {
      file: outputName,
      format: options.format,
      name: umdName,
      sourcemap: true,
      globals: {
        'react-hook-form': 'ReactHookForm',
      },
      exports: 'named',
    },
...
```

**Comment by TheDSCPL** [comment]
@bluebill1049 [There you go](https://github.com/TheDSCPL/resolvers/actions/runs/260663039)

**Comment by TheDSCPL** [comment]
@jorisre if you set ajv as an external it won't be bundled and hence not work on the browser

**Comment by bluebill1049** [comment]
hey @TheDSCPL you happy to pull that commit out from master? I will create a PR again from that commit until we find a fix for it.

**Comment by TheDSCPL** [comment]
I'll make a new PR with the new changes [that passed the CI build](https://github.com/TheDSCPL/resolvers/actions/runs/260663039).
